---
title: "MBQN vignette"
#date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MBQN Package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The MBQN package supplies the mean/median-balanced quantile normalization for preprocessing omics or other matrix-like organized data with intensity values distorted by global shifts of mean and scatter of columns [1]. The mean intensity of rank invariant (RI) or nearly rank invariant (NRI) features, e.g. features that have always largest intensity across samples/columns or are otherwise intensity-separated from the other features, is balanced before application of conventional quantile normalization in order to reduce systematics in downstream analysis. 

## Installation
To install this package from Github, type in R (version >="3.5"): 
```{r gh-installation, eval = FALSE}
# install.packages("devtools")
devtools::install_github("arianeschad/mbqn")
```
or 
```{r installation, eval = FALSE}
# install.packages("githubinstall")
githubinstall::githubinstall("mbqn")
```

To install this package from Bioconductor, type in R (version >="3.5"): 
```{r bc-installation, eval = FALSE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
BiocManager::install("mbqn")
```

## Dependencies
The package depends on the `limma` package [2] for computation of the conventional quantile normalization, available at Bioconductor. 
Alternatively, `normalize.quantiles()` from the package preprocessCore [3], available from https://github.com/bmbolstad/preprocessCore can be used.

Collecting data from PRIDE experiments in `mbqnExample()` requires the R package rpx [4] available from https://github.com/lgatto/rpx or via Bioconductor.

Additional packages needed to run MBQN examples, R package `filestrings`. 

After installation, check the `?mbqnDemo` and the `?mbqnExample` help content for full working examples.

## Basic Usage
The package provides two basic functions: `mbqn()` applies quantile normalization or mean/median-balanced quantile normalization to a matrix. `mbqnNRI()` applies quantile normalization and mean/median-balanced quantile normalization only to selected nearly rank invariant and rank invariant features, specified by a threshold or manually. The input matrix may contain NAs. To run one of these functions you will need to provide an input matrix similar to the data matrix in `mbqnDemo` or `mbqnExample`. The argument `FUN` is used to select between classical quantile normalization (default), and mean or median balanced quantile normalization. The function `mbqnGetNRIfeatures()` and `mbqnPlotRI()` can be used to check a data matrix for rank or nearly rank invariant features. They provide a list of potential RI/NRI features together with their rank invariance frequency and a graphical output. 

## Examples
Example 1: Generate a distorted omics-like matrix of log2-transformed intensities with missing values and a single rank invariant feature:
```{r example1, eval = TRUE}
## basic example
library("MBQN")
mtx <- mbqnSimuData("omics.dep")
mtx <- mbqnSimuDistortion(mtx)$x.mod
```
```{r figure1, fig.height = 5, fig.width = 6, fig.align = "center", fig.cap = "Fig. 1 Boxplot of the unnormalized, distorted intensity data matrix. The first feature is an RI feature (red line). It has maximum intensity for each sample!"}
plot.new()
mbqnBoxplot(mtx, irow = 1, main = "unnormalized")
```
<br />
Apply check for rank invariant (RI) or nearly rank invariant (NRI) features to the data matrix and visualize result:
```{r, eval = TRUE}
ri.obj <- mbqnGetNRIfeatures(mtx, low_thr = 0.5)
```
```{r figure2, fig.height = 3, fig.width = 6, fig.align = "center", fig.cap = "Fig. 2 Correctly detected and identified RI feature with a data coverage of 100% across samples."}
plot.new()
mbqnPlotRI(ri.obj)
```
<br />
Apply quantile normalization with and without balancing the RI feature and compare the intensity features: 
```{r figure3, fig.height = 5, fig.width = 6, fig.align = 'center', fig.cap = "Fig. 3 Quantile normalized intensities with balanced and unbalanced normalized RI feature. Classical quantile normalization suppresses any intensity variation of the RI feature, while the MBQN preserves its variation while reducing systematic batch effects!"}
plot.new()
mtx.norm <- mbqnNRI(x = mtx, FUN = median, verbose = FALSE) # MBQN
mtx.qn <- mbqnNRI(x = mtx, FUN = NULL, verbose = FALSE) # QN
mbqnBoxplot(mtx.norm, irow = ri.obj$ip, vals = data.frame(QN = mtx.qn[ri.obj$ip,]), main = "normalized")
```

<br /><br />
Example 2: This example downloads an LFQ intensity dataset from the PRIDE repository, normalizes the data, identifies RI/NRI features, and give graphical output. One can choose between four data sets. By default data files are stored in the current working directory currentdir/PXDxxx.

```{r example2, eval = FALSE}
## Normalize LFQ intensity data from the PRIDE repository
mbqnExample(which.example = 1)
```

In order to run `mbqnExample()`, the respective proteinGroups.txt file must be first downloaded from PRIDE, e.g. with `MBQN:::get_pxdfile()` to a user-specified directory or you can directly download the data by running `mbqnExample()` to currentdir/PXDxxx. 

## Figures
Figures created with MBQN are saved as pdf in the current working directory.

## Documentation
To view the documentation for MBQN, type the following command in R `??MBQN`.

## References
[1] A. Schad and C. Kreutz, MBQN: an R/Bioconductor package for mean/ median-balanced quantile normalization. In prep. 2019 <br/>
[2] Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research 43(7), e47. <br/>
[3] Ben Bolstad (2018). preprocessCore: A collection of pre-processing functions. R package version 1.44.0. https://github.com/bmbolstad/preprocessCore. <br/>
[4] Laurent Gatto (2019). rpx: R Interface to the ProteomeXchange Repository. R package version 1.18.1. https://github.com/lgatto/rpx. 
